src/main/java/com/avispl/symphony/dal/device/core510i/QRCCommunicator.java

```diff
@@ -0,0 +1,440 @@
+	 * @return String[] This returns the reply received from the device.
+	 */
+	private String[] send(String data, boolean retryOnError) throws Exception {
+		try {
+			if (!this.isChannelConnected()) {
+				this.createChannel();
+				this.status.setLastTimestamp(System.currentTimeMillis());
+				this.status.setConnectionState(ConnectionState.Connected);
+				this.status.setLastError(null);
+			}
+
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug("Sending: " + data + " to: " + this.host + " port: " + this.port);
+			}
+
+			String[] response = this.internalSend(data);
+
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug("Received response: " + Arrays.toString(response) + " from: " + this.host + " port: " + this.port);
+			}
+
+			if (this.logger.isTraceEnabled()) {
+				this.logger.trace("Received response: " + Arrays.toString(response) + " from: " + this.host + " port: " + this.port);
+			}
+
+			this.status.setLastTimestamp(System.currentTimeMillis());
+			return response;
+		} catch (CommandFailureException ex1) {
+			if (this.logger.isErrorEnabled()) {
+				this.logger.error("Command failed " + data + " to: " + this.host + " port: " + this.port + " connection state: " + this.status.getConnectionState(), ex1);
+			}
+
+			this.status.setLastTimestamp(System.currentTimeMillis());
+			throw ex1;
+		} catch (SocketTimeoutException ex2) {
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug("Time out while sending command: " + data + " to: " + this.host + " port: " + this.port + " connection state: " + this.status.getConnectionState() + " error: ", ex2);
+			}
+
+			this.status.setLastError(ex2);
+			this.status.setConnectionState(ConnectionState.Unknown);
+			this.destroyChannel();
+
+			if (retryOnError) {
+				return this.send(data, false);
+			} else {
+				throw ex2;
+			}
+		} catch (Exception ex3) {
+			if (this.logger.isErrorEnabled()) {
+				this.logger.error("Error sending command: " + data + " to: " + this.host + " port: " + this.port + " connection state: " + this.status.getConnectionState() + " error: ", ex3);
+			}
+
+			this.status.setLastError(ex3);
+			this.status.setConnectionState(ConnectionState.Failed);
+			this.destroyChannel();
+
+			if (retryOnError) {
+				return this.send(data, false);
+			} else {
+				throw ex3;
+			}
+		}
+	}
+
+	private String[] internalSend(String outputData) throws IOException {
+		this.write(outputData);
+		return this.read(outputData, this.socket.getInputStream());
+	}
+
+	private void write(String outputData) throws IOException {
+		if (this.socket == null) {
+			throw new IllegalStateException("Socket connection was not established. Please check target host availability and credentials.");
+		} else {
+			OutputStream os = this.socket.getOutputStream();
+			OutputStreamWriter osw = new OutputStreamWriter(os);
+			osw.write(outputData);
+			osw.flush();
+		}
+	}
+
+	private String[] read(String command, InputStream in) throws IOException {
```

Comment:
javadocs for private methods won't hurt also, since it will useful for someone seeing this code and would remove the necessity (partially) of reading through the code to get the idea of the functionality

Reply:
Ok thank you, I will add javadocs for it